libbpf-tools/schedtimes.bpf.c

```diff
@@ -0,0 +1,200 @@
+    
+}
+
+static bool allow_record(struct task_struct *p){
+    u32 pid, tgid, ppid;
+    pid = BPF_CORE_READ(p, pid);
+    tgid = BPF_CORE_READ(p, tgid);
+    ppid = BPF_CORE_READ(p, real_parent, tgid);
+    if( pid == 0 )
+        return false;
+    if (target_pid && pid != target_pid)
+        return false;
+    else if(target_tgid && tgid != target_tgid)
+        return false;
+    else if (target_ppid && ppid != target_ppid)
+        return false;
+    return true;
+}
+
+static int handle_switch(struct task_struct *prev, struct task_struct *next)
+{
+    u32 prev_pid;
+    u32 next_pid;
+    u64 delta;
+    u32 state = get_task_state(prev);
+    struct state_ts_t state_ts, *state_tsp;
+
+    __builtin_memset(&state_ts, 0, sizeof(state_ts));
+    state_ts.ts = bpf_ktime_get_ns();
+    
+    //prev_pid sched out
+    prev_pid = BPF_CORE_READ(prev, pid);
+    if(allow_record(prev)){
+        //caculate prev run_time
+        state_tsp = bpf_map_lookup_elem(&start, &prev_pid); 
+        if(state_tsp){
+            delta = state_ts.ts - state_tsp->ts;
+            update_times(prev, delta, STATE_RUNNING); 
+            bpf_map_delete_elem(&start, &prev_pid);
+        }
+
+        //update start time
+        if(state == TASK_RUNNING){
+            state_ts.state = STATE_QUEUED;
+        }else if(state & TASK_INTERRUPTIBLE){
+            state_ts.state = STATE_SLEEPING;
+        }else if((state & TASK_UNINTERRUPTIBLE) && !(state & TASK_NOLOAD)){
+            state_ts.state = STATE_BLOCKED;
+        }
+        bpf_map_update_elem(&start, &prev_pid, &state_ts, 0);
+
+    }
+    //next_pid sched in
+    next_pid = BPF_CORE_READ(next, pid);
+    if(allow_record(next)){
+        //calculate queue time
+        state_tsp = bpf_map_lookup_elem(&start, &next_pid);
+        if(state_tsp && state_tsp->state == STATE_QUEUED){
+            delta = state_ts.ts - state_tsp->ts;
+            update_times(next, delta, STATE_QUEUED);
+            bpf_map_delete_elem(&start, &next_pid);
+        }
+
+        //update start time
+        //state_ts.state = STATE_RUNNING;
+        bpf_map_update_elem(&start, &next_pid, &state_ts, 0);
+    }
+    return 0;
+}
+
+static int wakeup(struct task_struct *p)
+{
+    u32 pid; 
+    u64 delta;
+    struct state_ts_t state_ts, *state_tsp;
+    __builtin_memset(&state_ts, 0, sizeof(state_ts));
+    state_ts.ts = bpf_ktime_get_ns();
+
+    pid = BPF_CORE_READ(p, pid);
+    if(!allow_record(p))
+        return 0;
+    //caculate sleep/block time
+    state_tsp = bpf_map_lookup_elem(&start, &pid);
+    if(state_tsp){
+        delta = state_ts.ts - state_tsp->ts;
+        if(state_tsp->state == STATE_SLEEPING)
+            update_times(p, delta, STATE_SLEEPING);
+        if(state_tsp->state == STATE_BLOCKED)
+            update_times(p, delta, STATE_BLOCKED);
+        bpf_map_delete_elem(&start, &pid);
+    }
+    //update queue time
+    state_ts.state = STATE_QUEUED;
+    bpf_map_update_elem(&start, &pid, &state_ts, BPF_ANY);
```

Comment:
We have bpf_map_delete_elem above followed by bpf_map_update_elem. There is no need for this. You can directly modify state_tsp contents if not null.